narwhal/primary/src/proposer.rs

```diff
@@ -512,33 +515,38 @@ impl Proposer {
 
                 Some((commit_round, commit_headers)) = self.rx_committed_own_headers.recv() => {
                     // Remove committed headers from the list of pending
-                    for round in &commit_headers {
-                        self.proposed_headers.remove(round);
+                    let mut max_committed_round = 0;
+                    for round in commit_headers {
+                        max_committed_round = max_committed_round.max(round);
+                        let Some(_) = self.proposed_headers.remove(&round) else {
+                            info!("Own committed header not found at round {round}, probably because of restarts.");
+                            continue;
+                        };
                     }
 
-                    // Now for any round much below the current commit round we re-insert
+                    // Now for any round below the current commit round we re-insert
                     // the batches into the digests we need to send, effectively re-sending
-                    // them
+                    // them in FIFO order.
                     let mut digests_to_resend = Vec::new();
                     let mut retransmit_rounds = Vec::new();
 
                     // Iterate in order of rounds of our own headers.
-                    for (round_header, header) in &mut self.proposed_headers {
-                        // If we are within 2 rounds of the commit we break
-                        if round_header + 2 >= commit_round {
+                    for (header_round, (_header, included_digests)) in &mut self.proposed_headers {
+                        // Stop once we have processed headers at and below last committed round.
+                        if *header_round > max_committed_round {
                             break;
                         }
-
-                        digests_to_resend.extend(
-                            header.payload.iter().map(|(digest, (worker, created_at))| (*digest, *worker, *created_at))
-                        );
-                        retransmit_rounds.push(*round_header);
+                        // Add payloads from oldest to newest.
+                        digests_to_resend.append(included_digests);
+                        retransmit_rounds.push(*header_round);
                     }
 
                     if !retransmit_rounds.is_empty() {
                         // Add the digests to retransmit to the digests for the next header
-                        digests_to_resend.extend(&self.digests);
-                        self.digests = digests_to_resend;
+                        let num_to_resend = digests_to_resend.len();
+                        for digest in digests_to_resend.into_iter().rev() {
+                            self.digests.push_front(digest);
+                        }
```

Comment:
Is the intention to prioritise on the final `self.digests` list the newer batches first? If yes, then doing a `rev` on the `digests_to_resend` and a `push_front` will result in adding the elements in an order of `older --> newer`  . So either remove the `rev` or `push_back` ?